catalog/spatial/csv/spatial-csv-transformer/src/main/java/org/codice/ddf/spatial/ogc/csv/catalog/transformer/MetacardIterator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+package org.codice.ddf.spatial.ogc.csv.catalog.transformer;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import ddf.catalog.data.Attribute;
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.Metacard;
+
+/**
+ * An implementation of java.util.Iterator which iterates over Metacard attribute values.
+ *
+ * @see java.util.Iterator
+ */
+class MetacardIterator implements Iterator<Serializable> {
+    private List<AttributeDescriptor> attributeDescriptorList;
+
+    private Metacard metacard;
+
+    private int index;
+
+    /**
+     * @param metacard                the metacard to be iterated over.
+     * @param attributeDescriptorList the list of attributeDescriptors used to determine which
+     *                                metacard attributes to return.
+     */
+    MetacardIterator(final Metacard metacard,
+            final List<AttributeDescriptor> attributeDescriptorList) {
+        this.metacard = metacard;
+        this.attributeDescriptorList = Collections.unmodifiableList(attributeDescriptorList);
+        this.index = 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasNext() {
+        return this.attributeDescriptorList.size() > index;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Serializable next() {
+        if (!this.hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        AttributeDescriptor attributeDescriptor = this.attributeDescriptorList.get(index);
+        Attribute attribute = metacard.getAttribute(attributeDescriptor.getName());
+        index++;
+
+        if (attribute != null) {
+            return attribute.getValue();
+        }
+
+        return "";
```

Comment:
'attributeDescriptorList' is going to contain the union of all the AttributeDescriptors from a returned query because the output CSV has to contain a column for each one. `metacard.getAttribute("nitf.imageMode") is going to return an attribute for Nitf metacards, but not for FMV ones (for example). So we will frequently be getting null attributes returned and calling `.getValue()` on null is going to cause an NPE. The CSVPrinter seems to handle null and "" the same way, so I'm not sure that `return null` is better than `return ""` in the case where the attribute object is null.

In short, if the attribute's value is null then null will be returned.  If the attribute doesn't exist in the metacard, you'll get back "". And for the purpose of generating a CSV file, they're equivalent.